Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Utilities/JetBrain/JetBrains.Annotations.cs

```diff
@@ -0,0 +1,1072 @@
+﻿/* MIT License
+Copyright (c) 2016 JetBrains http://www.jetbrains.com
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE. */
+
+/*
+    This file is downloaded from https://github.com/bepost/jetbrains-annotations-internal    
+*/
+
+using System;
+
+#pragma warning disable 1591
+// ReSharper disable UnusedMember.Global
+// ReSharper disable MemberCanBePrivate.Global
+// ReSharper disable UnusedAutoPropertyAccessor.Global
+// ReSharper disable IntroduceOptionalParameters.Global
+// ReSharper disable MemberCanBeProtected.Global
+// ReSharper disable InconsistentNaming
+// ReSharper disable UnusedParameter.Local
+// ReSharper disable InconsistentNaming
+// ReSharper disable CheckNamespace
+// ReSharper disable ArrangeThisQualifier
+// ReSharper disable CommentTypo
+// ReSharper disable IdentifierTypo
+
+namespace JetBrains.Annotations
```

Comment:
Hmm. Include it or not include. ❓ 🤔

Reply:
I have no problem using this, since the Entity Framework Core uses it as well.

We use it only for function annotations.

Reply:
Well, currently there's no static analysis doing something with the attributes. So it has very little value.

Reply:
We can see if it will really need to be removed.
I'm using annotation in many places.

I will see this and the possibility to remove it.

Reply:
I'd like do hear also your opinion why to use it or not. And what it brings. I want some discussion with you guys. 

I don't think there's only one correct answer for this.

Reply:
**My intention** is to make markings on the function so that you do not have to do any over-treatment.

In this way we will ensure that the object will always be null. 
Hence the use of the "NotNullAttribute".
In the interfaces I use to have more security.

Practical example:

```csharp
public interface IFbUpdateSqlGenerator: IUpdateSqlGenerator
{
	ResultSetMapping AppendBlockInsertOperation (
		[NotNull] StringBuilder commandStringBuilder,
		[NotNull] StringBuilder headBlockStringBuilder,
		[NotNull] IReadOnlyList <ModificationCommand> modificationCommands,
		int commandPosition);

	ResultSetMapping AppendBlockUpdateOperation (
		[NotNull] StringBuilder commandStringBuilder,
		[NotNull] StringBuilder headBlockStringBuilder,
		[NotNull] IReadOnlyList <ModificationCommand> modificationCommands,
		int commandPosition);

	ResultSetMapping AppendBlockDeleteOperation (
	  [NotNull] StringBuilder commandStringBuilder,
	  [NotNull] IReadOnlyList <ModificationCommand> modificationCommands,
	  int commandPosition);
}
```

Reply:
Where's the security?

I can do:
```
interface IFoo
{
	void FooBar([NotNull] StringBuilder foobar);
}

class Foo : IFoo
{
	public void FooBar([NotNull] StringBuilder foobar)
	{ }
}
```

And call it `new Foo().FooBar(null);` it will compile and run just fine. I would be a nice touch to have it in IntelliSense for example, kind of at least something. But unless I'll use some post-build process i.e. _PostSharp_ to inject some validation, I don't see it. Am I missing something?